README.md
# btrfs-cleaner
Do periodic maintenance on a BTRFS filesystem.  Yes, I'd call running this everynight "periodic."

##

    usage: btrfs-cleaner [-h] [-F FILLED] [-b] [-d] [-n] [-o OFILE] [-s] [-v]
                         [--version]
                         [mountpoint [mountpoint ...]]

    Scrub btrfs filesystems in background

    positional arguments:
      mountpoint            filesystems to scrub (default: [])

    optional arguments:
      -h, --help            show this help message and exit
      -F FILLED, --filled FILLED
                            list of balance points (default: 1 2 3 5 7 10 15 20 30
                            40 50 60 70 80 90 100)
      -b, --balance         balance filesystem (default: False)
      -d, --defrag          defragment files (default: False)
      -n, --dont            print what would be done (default: False)
      -o OFILE, --out OFILE
                            output here if not stdout (default: None)
      -s, --scrub           recalculate checksum of chunks (default: False)
      -v, --verbose         show command being run (default: False)
      --version             btrfs-cleaner v0.0.7

    If no action (defrag, balance, or scrub) is specified then all three actions
    will be performed.


ChangeLog
* Fri Oct 25 2019 Tommy Reynolds
- (9ef719d) Bump version (HEAD, master)

* Fri Oct 25 2019 Tommy Reynolds
- (acb28d3) Right-align test times and duration

* Fri Jun 14 2019 Tommy Reynolds
- (a950504) Imported (origin/master, origin/HEAD)

* Sat Jun 1 2019 Tommy Reynolds
- (3224f14) Use the socially-acceptible #!/usr/bin/env python2 shebang.

* Thu Mar 28 2019 Tommy Reynolds
- (484b425) Explain ourselves a bit.

* Thu Mar 28 2019 Tommy Reynolds
- (e100e1b) Merge branch 'master' of ssh://gmbh.dontconfuse.us:2223/home/vault/depot/btrfs-cleaner

* Thu Mar 28 2019 Tommy Reynolds
- (9b3030a) Initial commit

* Fri Mar 8 2019 Tommy Reynolds
- (e5062f7) Changes (tag: v0.0.8)

* Fri Mar 8 2019 Tommy Reynolds
- (bf29db2) Additions

* Fri Mar 8 2019 Tommy Reynolds
- (a5f9c7a) Sample cron(1) entry

* Fri Jan 4 2019 Tommy Reynolds
- (784e820) Show example actions.

* Fri Jan 4 2019 Tommy Reynolds
- (dd2b61c) Write output to the 'ofile', everywhere. (tag: v0.0.7)

* Fri Jan 4 2019 Tommy Reynolds
- (54937ab) Calculate processing times. (tag: v0.0.6)

* Fri Jan 4 2019 Tommy Reynolds
- (40070c4) Changed method name to be more accurate.

* Fri Jan 4 2019 Tommy Reynolds
- (55a6566) Show default values if '--help' is used.

* Fri Jan 4 2019 Tommy Reynolds
- (5414202) Show initial geometry; added 'df' info; prettier.

* Thu Jan 3 2019 Tommy Reynolds
- (eeeb315) Bumped (tag: v0.0.3)

* Thu Jan 3 2019 Tommy Reynolds
- (9344ed5) Dropped all the bother about crosschecking so much.

* Thu Jan 3 2019 Tommy Reynolds
- (e421f3d) Modernize bunchers

* Thu Jan 3 2019 Tommy Reynolds
- (5b515a7) Bump version.

* Thu Jan 3 2019 Tommy Reynolds
- (95d3547) Do steps in this order: scrub, balance, defrag

* Thu Jan 3 2019 Tommy Reynolds
- (a9911e2) Corrected error message not being written to stderr.

* Thu Jan 3 2019 Tommy Reynolds
- (035a683) Added balance points

* Wed Feb 28 2018 Tommy Reynolds
- (fa2217b) Basked of Ignorables.

* Wed Feb 28 2018 Tommy Reynolds
- (38c0f89) Do steps in a better sequence: scrub, defrag, balance.

* Sun Feb 25 2018 Tommy Reynolds
- (16b7c0d) Checkpoint

* Sat Feb 24 2018 Tommy Reynolds
- (ead6ebc) History

* Sat Feb 24 2018 Tommy Reynolds
- (82339ed) Name the runtime script.

* Sat Feb 24 2018 Tommy Reynolds
- (79c08d4) Mea culpa

* Sat Feb 24 2018 Tommy Reynolds
- (ff83ce8) Basket of Ignorables

* Sat Feb 24 2018 Tommy Reynolds
- (ddbf99b) Baseline



btrfscleaner/bunch.py
#!/usr/bin/env python2
# vim: noet sw=4 ts=4

import	sys
import	os

class	Bunch( dict ):

	def	__init__( self, **kwds ):
		super( Bunch, self ).__init__( **kwds )
		self.__dict__.update( **kwds )
		return

	def	__getattr__( self, name ):
		return self.__dict__.get( name, None )

	def	__setattr__( self, name, value ):
		self.__dict__[name] = value

	def	__getitem__( self, key ):
		return self.__dict__.get( key, None )

	def	__setitem__( self, key, value ):
		self.__dict__[ key ] = value
		return

	def	__delattr__( self, name ):
		if name in self.__dict__:
			del self.__dict__[name]
		return

	def	__iter__( self ):
		return iter( self.__dict__ )

	def	__repr__( self ):
		parts = [
			"'{0}': '{1}'".format( l, self.__dict__[l] ) for l in self.__dict__ if l[0] != '_'
		]
		return '{{ {0} }}'.format(
			'; '.join( parts )
		)

if __name__ == '__main__':
	b = Bunch(
		first = 'Tommy',
		last = 'Reynolds'
	)
	print 'Name: {0}, {1}, Should be none={2}'.format(
		b.last,
		b.first,
		b.no_such_value
	)
	print 'b={0}'.format( b )
	print 'last={0}'.format( b.get( 'last','BOO!' ) )
	print 'missing={0}'.format( b.get( ' *missing* ','BOO!' ) )
	print 'iter={0}'.format(
		[ x for x in b ]
	)
	print 'iter:'
	for i,key in enumerate( b ):
		print 'key[{0}]={1}'.format( i+1, key )
	exit( 0 )


btrfscleaner/__init__.py
#!/usr/bin/env python2
# vim: filetype=python noet sw=4 ts=4

import	argparse
import	bunch
import	datetime
import	os
import	shlex
import	subprocess
import	sys
import	syslog

try:
	from version import Version
except:
	Version = 'W.T.F.'

class	BtrfsCleaner( object ):

	def	indent( self, s ):
		return '    {0}'.format( s )

	def	log( self, s, pri = syslog.LOG_ERR ):
		if self.out:
			print >>self.out, self.indent( s )
		syslog.syslog( pri, s )
		return

	def	new_node( self, where = None, dev = None ):
		return bunch.Bunch( where = where, dev = dev )

	def	select( self ):
		#
		# Make a local copy of the active mount table, then reduce it to
		# a dict of BTRFS mountpoint names.
		#
		with open( '/proc/mounts' ) as f:
			mounts = map(
				str.split,
				f.readlines(),
			)
		mounts[:] = [
			m[1] for m in mounts if m[2] == 'btrfs'
		]
		# print 'mounts={0}'.format( mounts )
		return mounts

	def	__init__(
		self
	):
		# Get logging setup early
		syslog.openlog(
			'scrub-btrfs',
			syslog.LOG_PID,
			syslog.LOG_DAEMON,
		)
		self.out         = None
		# Record simple options
		self.opts        = bunch.Bunch()
		self.opts.dont   = False
		self.opts.filled = "1 2 3 5 7 10 15 20 30 40 50 60 70 80 90 100"
		return

	def	run( self, cmd, prolog = None ):
		cli = '{0} {1}'.format(
			'$' if os.getuid() else '#',
			' '.join( cmd )
		)
		err    = None
		if not prolog:
			prolog = list()
		elif type(prolog) != list:
			prolog = [ prolog ]
		output = prolog
		if self.opts.dont:
			output.append( cli )
		else:
			if self.opts.verbose:
				output.append( cli )
			try:
				s = subprocess.check_output(
					cmd,
					stderr = subprocess.STDOUT,
				)
				output.append( s )
			except subprocess.CalledProcessError, e:
				err = [ e.output ] +  [
					'Exit code {0}'.format( e.returncode ),
				]
			except Exception, e:
				print '*** EXCEPTION {0}'.format( e )
				raise e
		return output, err

	def	printLn( self, s = '' ):
		print >>self.out, s
		return

	def	show( self, output = None, err = None ):
		fmt = '{0:<2} {1}'
		if output:
			for part in output:
				for line in part.splitlines():
					self.printLn( self.indent( fmt.format( '', line ) ) )
		if err:
			for part in err:
				for line in part.splitlines():
					self.printLn( self.indent( fmt.format( '**', line ) ) )
		return

	def	report( self ):
		pass

	def	do_scrub( self, mp ):
		# Scrub it
		cmd = [
			'/sbin/btrfs',
			'scrub',
			'start',
			'-B',
			'-d',
			'-f',
			mp
		]
		output, err = self.run( cmd )
		self.show( output, err )
		return

	def	do_balance( self, mp ):
		# Balance it
		for fill in self.opts.filled:
			cmd = [
				'/sbin/btrfs',
				'balance',
				'start',
				'-dusage={0}'.format( fill ),
#				'-dlimit=32',
				'-musage={0}'.format( fill ),
#				'-mlimit=32',
				mp
			]
			output, err = self.run(
				cmd,
				prolog = 'Balancing {0}% usage'.format( fill ),
			)
			self.show( output, err )
		return

	def	do_defrag( self, mp ):
		# Defrag
		cmd = [
			'/sbin/btrfs',
			'filesystem',
			'defragment',
			'-f',
			'-r',
			mp
		]
		output, err = self.run( cmd )
		self.show( output, err )
		return

	def	do_df( self, mp ):
		# Defrag
		cmd = [
			'/sbin/btrfs',
			'filesystem',
			'df',
			'--iec',
			mp
		]
		output, err = self.run( cmd )
		self.show( output, err )
		return

	def	do_du( self, mp = None ):
		if not mp:
			mp = self.opts.filesystems
		cmd = [
			'/sbin/btrfs',
			'filesystem',
			'du',
			'-s',
			'--iec',
		] + sorted( mp )
		output, err = self.run( cmd )
		self.show( output, err )
		return

	def	section( self, title, banner = None, step = True ):
		if step:
			self.step += 1
			leadin = '{0}. '.format( self.step )
		else:
			leadin = ''
			self.step = 0
		print >>self.out
		print >>self.out, '{0}{1}'.format( leadin, title )
		if banner:
			print >>self.out, '{0}{1}'.format( '', banner * len( title ) )
		print >>self.out
		return

	def	main( self ):
		prog = os.path.splitext(
			os.path.basename( sys.argv[0] )
		)[0]
		if prog == '__init__':
			prog = 'btrfs-cleaner-bis'
		p = argparse.ArgumentParser(
			prog        = prog,
			description = '''\
				Scrub btrfs filesystems in background
			''',
			formatter_class = argparse.ArgumentDefaultsHelpFormatter,
			epilog = '''\
				If no action (defrag, balance, or scrub) is specified then
				all three actions will be performed.
			'''
		)
		p.add_argument(
			'-F',
			'--filled',
			dest    ='filled',
			help    = 'list of balance points',
			default = self.opts.filled,
		)
		p.add_argument(
			'-b',
			'--balance',
			dest   = 'balance',
			action = 'store_true',
			help   = 'balance filesystem'
		)
		p.add_argument(
			'-d',
			'--defrag',
			dest   = 'defrag',
			action = 'store_true',
			help   = 'defragment files',
		)
		p.add_argument(
			'-n',
			'--dont',
			dest    = 'dont',
			default = False,
			action  = 'store_true',
			help    = 'print what would be done'
		)
		p.add_argument(
			'-o',
			'--out',
			dest = 'ofile',
			help = 'output here if not stdout',
		)
		p.add_argument(
			'-s',
			'--scrub',
			dest   = 'scrub',
			action = 'store_true',
			help   = 'recalculate checksum of chunks',
		)
		p.add_argument(
			'-v',
			'--verbose',
			dest   = 'verbose',
			action = 'store_true',
			help   = 'show command being run',
		)
		p.add_argument(
			'--version',
			action  = 'version',
			version = Version,
			help    = '{0} v{1}'.format( prog, Version )
		)
		p.add_argument(
			'filesystems',
			metavar = 'mountpoint',
			nargs = '*',
			default = [],
			help = 'filesystems to scrub'
		)
		self.opts = p.parse_args()
		if os.getuid() != 0:
			print >>sys.stderr, 'Not running as root; expect troubles.'
		#
		ofile = self.opts.ofile
		self.out = open( ofile, 'wt' ) if ofile else sys.stdout
		#
		active_mounts = self.select()
		if len( self.opts.filesystems ) == 0:
			self.opts.filesystems = sorted( active_mounts )
		#
		if not any((
			self.opts.balance,
			self.opts.defrag,
			self.opts.scrub,
		)):
			self.opts.balance = True
			self.opts.defrag  = True
			self.opts.scrub   = True
		# For balancing, convert the space-delimited list of
		# usage thresholds into a list of numbers.
		if self.opts.balance:
			try:
				a = [
					int(x) for x in self.opts.filled.split()
				]
				self.opts.filled = a
			except:
				print >>sys.stderr, 'Balance points must be numeric: {0}'.format(
						self.opts.filled,
					)
				exit( 1 )
		#
		# Here we go
		#
		title = 'BTRFS Cleaning'
		self.printLn( title )
		self.printLn( '=' * len( title ) )
		self.printLn( '')
		# Make an establishing shot
		self.section( 'Filesystems Under Scrutiny', step = False )
		self.do_du()
		#
		uuids_already_processed = dict()
		for mp in sorted( self.opts.filesystems ):
			time_started = datetime.datetime.now()
			if mp not in active_mounts:
				print >>sys.stderr, 'Not a mount point: {0}'.format(
					mp
				)
				continue
			self.section(
				'Mountpoint "{0}"'.format( mp ),
				banner = '-',
				step = False,
			)
			# Make sure we can belive the filesystem metadata
			if self.opts.scrub:
				self.section( 'Scrubbing' )
				self.do_scrub( mp )
			# Be awesome
			self.section( 'Filesystem Topology' )
			self.do_df( mp )
			# Repack the filesystem to maximize free space
			if self.opts.balance:
				self.section( 'Balancing' )
				self.do_balance( mp )
			# Consolidate file disk usage
			if self.opts.defrag:
				self.section( 'Defragmenting (expect errors)' )
				self.do_defrag( mp )
			#
			self.section( 'Done' )
			time_ended = datetime.datetime.now()
			time_span  = time_ended - time_started
			fmt        = '{0:<8} {1:>27}'
			output     = [
				fmt.format( 'Ended',    str( time_ended )	),
				fmt.format( 'Started',  str( time_started )	),
				fmt.format( 'Duration', str( time_span )	),
			]
			self.show(
				output = output,
			)
		return 0

if __name__ == '__main__':
	exit( BtrfsCleaner().main() )
